### DISCUSSION TITLE
Delete history entry while browsing matches menu

### DISCUSSION THREAD

**Discussion by ardalonph** [junegunn/fzf#2800] [Ideas]
Firstly, hello and thank you for revolutionizing how I use the terminal.  I can no longer live without the productivity gains I get from FZF and my skill operating *nix (and even powershell) have improved dramatically. 

Is there any chance you would consider building functionality into FZF that would allow me to delete erroneous history as I'm browsing? Its just such a great way to browse history and retrieve commands it  might also be great to be able to remove the ones I come across that I know don't work

A use case:
- I type git clone x ... i've cloned this before, i know its in the history so ....
- I type ctrl-R the menu of fuzzy matches appears
- I see 3 commands with typos in them and the command I want
- before I choose the match I want, i arrow up to 1 offending command and i press delete
- the command disappears from the list (and underlying history) but the menu does not close giving me the opportunity to select more
- i arrow to the next bad commands and I press delete for each
- i choose what i came for and press enter

**Comment by junegunn** [comment]
I wouldn't add the feature to CTRL-R implementation this repository provides not to increase the maintenance burden. But if you decide to implement your own version of CTRL-R, I believe you can achieve that using `execute-silent` and/or `reload` bindings.

```sh
mkdir -p /tmp/test
cd /tmp/test
touch {1..100}
FZF_DEFAULT_COMMAND=ls fzf --bind 'ctrl-d:reload(rm {}; $FZF_DEFAULT_COMMAND)' --header 'ctrl-d to delete the file'
```

**Reply by cohml** [reply]
@junegunn Your example shows how to rerun the default command. How would would trigger the rerunning of zsh's `fzf-history-widget` in a similar fashion?

For context: I'm trying to write some logic which, via a keybinding in `CTRL-R`, will delete the selected entry from my history. The deleting part seems to be working, but once deleted I'd like the fzf buffer contents to reload, equivalent to a fresh press of `CTRL-R`.

There is no convenient environment variable that I'm aware of which I could exploit for this, only a widget (for zsh anyway), but I've found that those are complicated to directly execute via fzf keybindings.

Any advice?

@ardalonph If you ever for something end-to-end working for using fzf to prune your history in the way you described above, I'd love to know how you did it. Bonus points for minimally viable code snippets that I can reuse :) Extra bonus points if it's zsh!

**Reply by LangLangBart** [reply]
> reload

The following setup allows you to reload the history using `fc -pa $HISTFILE;fc -rl 1 | awk …`. It's important to note that this method only displays history events globally. Therefore, it requires a `start` event to ensure that it does not display the current terminal session history, but only reads the history from the history file itself.

The `sed -i '' '$d' $HISTFILE` part is simply a demonstration that the last line is being deleted from the history file, and the reloading process effectively reflects these changes. The empty strings might be only necessary for macOS users.

> [!CAUTION]
> Modifying the history with `sed -i` is not approved by the maintainers of `zsh`. Refer to the following link for more information:
> [Re: How to (properly) remove the last entry from history with command_not_found_handler ](https://zsh.org/mla/users/2014/msg00714.html)


For additional explanation of the other flags, please refer to the `man fzf` manual page.

```bash
# location of my history file
export HISTFILE="$HOME/.dotfiles/zsh/.zsh_history"

export FZF_CTRL_R_OPTS="$(
	cat <<'FZF_FTW'
--bind "ctrl-d:execute(sed -i '' '$d' $HISTFILE)+reload:fc -pa $HISTFILE; fc -rl 1 |
	awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, \"\", cmd); if (!seen[cmd]++) print $0 }'"
--bind "start:reload:fc -pa $HISTFILE; fc -rl 1 |
	awk '{ cmd=$0; sub(/^[ \t]*[0-9]+\**[ \t]+/, \"\", cmd); if (!seen[cmd]++) print $0 }'"
--header 'enter select · ^d remove latest'
--height 100%
--preview-window "hidden:down:border-top:wrap:<70(hidden)"
--preview "bat --plain --language sh <<<{2..}"
--prompt " History > "
--with-nth 2..
FZF_FTW
)"
```

> [!TIP]
> The use of `awk …` may not be necessary if you set some options in your `.zshrc` file to prevent duplicates from being written to your history, refer to: [zsh: 16 Options ](https://zsh.sourceforge.io/Doc/Release/Options.html).
> ```bash
> # Exemplary settings 
> setopt HIST_EXPIRE_DUPS_FIRST # Expire duplicate entries first when trimming history.
> setopt HIST_IGNORE_DUPS       # Don't record an entry that was just recorded again.
> setopt HIST_IGNORE_ALL_DUPS   # Delete old recorded entry if new entry is a duplicate.
> setopt HIST_FIND_NO_DUPS      # Do not display a line previously found.
> setopt HIST_SAVE_NO_DUPS      # Don't write duplicate entries in the history file.
> setopt HIST_REDUCE_BLANKS     # Remove superfluous blanks before recording entry.
> ```

---

EDIT: Updated version with proper delete mechanic [Issue #3522](https://github.com/junegunn/fzf/issues/3522#issuecomment-1872415948)
